### ISSUE TITLE
junos_config broken configurations and idempotency

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29523]
From @Logan2211 on 2016-11-03T19:38:50Z
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
junos_config

##### ANSIBLE VERSION
```
ansible 2.2.0.0
  config file = /home/centos/provisioning-metal/ansible.cfg
  configured module search path = ['ephemeral_roles/plugins/library']
```

##### CONFIGURATION
Roles, plugins, inventory paths.

##### OS / ENVIRONMENT

CentOS 7 managing a Juniper EX3300 stack running 12.3R8.7

##### SUMMARY
The junos_config deploys broken/inconsistent configurations every time I run it.

##### STEPS TO REPRODUCE

Starting with blank port configuration:
```
logan@stack01.rack92> show configuration interfaces ae11
{master:0}
logan@stack01.rack92> show configuration interfaces ge-0/0/6
{master:0}
logan@stack01.rack92> show configuration interfaces ge-1/0/6
{master:0}
```

The playbook used is the following:
```
- debug:
    msg: "{{ switch_commands | default([]) }}"
  when: "{{ not switch_commands_fact | skipped }}"

- name: Apply JunOS configuration
  local_action:
    module: junos_config
    lines: "{{ switch_commands | default([]) }}"
    provider: "{{ hostvars[switch_port_hostname][hostvars[switch_port_hostname]['network_provider']] }}"
  when: "{{ not switch_commands_fact | skipped }}"
```

First run of playbook:
```
TASK [debug] *******************************************************************
ok: [lsn-mc1002] => {
    "msg": [
        "set interfaces vlan.2003 family inet address 10.3.8.1/24",
        "set vlans ANSIBLE-MANAGEMENT vlan-id 2003",
        "set vlans ANSIBLE-MANAGEMENT l3-interface vlan.2003",
        "delete interfaces ae11",
        "set interfaces ae11 description \"lsn-mc1002\"",
        "set interfaces ae11 aggregated-ether-options lacp active",
        "set interfaces ae11.0 family ethernet-switching port-mode access",
        "set interfaces ae11.0 family ethernet-switching vlan members ANSIBLE-MANAGEMENT"
    ]
}

TASK [Apply JunOS configuration] ***********************************************
changed: [lsn-mc1002 -> localhost]

TASK [debug] *******************************************************************
ok: [lsn-mc1002] => {
    "msg": [
        "set interfaces vlan.2003 family inet address 10.3.8.1/24",
        "set vlans ANSIBLE-MANAGEMENT vlan-id 2003",
        "set vlans ANSIBLE-MANAGEMENT l3-interface vlan.2003",
        "delete interfaces ge-0/0/6",
        "set interfaces ge-0/0/6 description \"lsn-mc1002\"",
        "set interfaces ge-0/0/6 ether-options 802.3ad ae11"
    ]
}

TASK [Apply JunOS configuration] ***********************************************
changed: [lsn-mc1002 -> localhost]

TASK [debug] *******************************************************************
ok: [lsn-mc1002] => {
    "msg": [
        "set interfaces vlan.2003 family inet address 10.3.8.1/24",
        "set vlans ANSIBLE-MANAGEMENT vlan-id 2003",
        "set vlans ANSIBLE-MANAGEMENT l3-interface vlan.2003",
        "delete interfaces ge-1/0/6",
        "set interfaces ge-1/0/6 description \"lsn-mc1002\"",
        "set interfaces ge-1/0/6 ether-options 802.3ad ae11"
    ]
}

TASK [Apply JunOS configuration] ***********************************************
changed: [lsn-mc1002 -> localhost]
```

Resulting configuration from first run looks good:
```
logan@stack01.rack92> show configuration interfaces ae11
description lsn-mc1002;
aggregated-ether-options {
    lacp {
        active;
    }
}
unit 0 {
    family ethernet-switching {
        port-mode access;
        vlan {
            members ANSIBLE-MANAGEMENT;
        }
    }
}
{master:0}
logan@stack01.rack92> show configuration interfaces ge-0/0/6
description lsn-mc1002;
ether-options {
    802.3ad ae11;
}
{master:0}
logan@stack01.rack92> show configuration interfaces ge-1/0/6
description lsn-mc1002;
ether-options {
    802.3ad ae11;
}
{master:0}
```

Second run. Ansible output is identical so I'll refrain from pasting it again.
```
logan@stack01.rack92> show configuration interfaces ae11
description lsn-mc1002;
unit 0 {
    family ethernet-switching {
        port-mode access;
        vlan {
            members ANSIBLE-MANAGEMENT;
        }
    }
}
{master:0}
logan@stack01.rack92> show configuration interfaces ge-0/0/6
description lsn-mc1002;
{master:0}
logan@stack01.rack92> show configuration interfaces ge-1/0/6
description lsn-mc1002;
{master:0}
```

3rd run:
```
logan@stack01.rack92> show configuration interfaces ae11
description lsn-mc1002;
aggregated-ether-options {
    lacp {
        active;
    }
}
unit 0 {
    family ethernet-switching {
        port-mode access;
        vlan {
            members ANSIBLE-MANAGEMENT;
        }
    }
}
{master:0}
logan@stack01.rack92> show configuration interfaces ge-0/0/6
description lsn-mc1002;
ether-options {
    802.3ad ae11;
}
{master:0}
logan@stack01.rack92> show configuration interfaces ge-1/0/6
description lsn-mc1002;
ether-options {
    802.3ad ae11;
}
{master:0}
```

Subsequent runs oscillate back and forth between the broken and proper configurations.
Copied from original issue: ansible/ansible-modules-core#5483

**Comment by ansibot** [comment]
From @ansibot on 2016-11-03T19:38:50Z



@privateip, @gundalow, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @Logan2211 on 2016-11-03T19:38:50Z

Maybe related to #3984 and #4007?

**Comment by ansibot** [comment]
From @ganeshrn on 2016-11-03T19:38:50Z

I figured out a few issues in code here https://github.com/ansible/ansible-modules-core/blob/devel/network/junos/junos_config.py#L212-L229

The config received from device (using 'get_config') is in text format and is converted to 'set' format in 'config_to_commands' function. The diff of config from playbook and one received from device fails for below cases during second run.
```
set interfaces ae11 description \"lsn-mc1002\"  ---> playbook config
set interfaces ae11 description lsn-mc1002     ---> config received from device and converted to set

set interfaces ae11.0 family ethernet-switching port-mode access ---> playbook config
set interfaces ae11 unit 0 family ethernet-switching port-mode access ---> config received from device and converted to set
```
Hence above playbook config is added in 'updates' list during second run and applied on device which is not expected.

@Logan2211 Can you please change the playbook config in below format and check if issue is seen
```
set interfaces ae11 description lsn-mc1002
set interfaces ae11 aggregated-ether-options lacp active
set interfaces ae11 unit 0 family ethernet-switching port-mode access
set interfaces ae11 unit 0 family ethernet-switching vlan members ANSIBLE-MANAGEMENT
```
Similarly for other interfaces
also remove the 'delete interface <-interface- name->' config in playbook as deleting and creating interface in same playbook won't guarantee idempotency.

@gundalow @privateip Please correct me if I am wrong.

**Comment by ansibot** [comment]
cc @rcarrillocruz @trishnaguha
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @kedarX
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ganeshrn** [comment]
needs_info

**Comment by ganeshrn** [comment]
As the issue is reported on a relatively older version, please retry on the latest release of Ansible version

**Comment by ansibot** [comment]
@Logan2211 This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@Logan2211 You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
